components/script/dom/htmlimageelement.rs

```diff
@@ -871,6 +917,175 @@ impl HTMLImageElement {
         ScriptThread::await_stable_state(Microtask::ImageElement(task));
     }
 
+    /// <https://html.spec.whatwg.org/multipage/#img-environment-changes>
+    pub fn react_to_environment_changes(&self) {
+        // Step 1
+        let task = ImageElementMicrotask::EnvironmentChangesTask {
+            elem: DomRoot::from_ref(self),
+            generation: self.generation.get(),
+        };
+        ScriptThread::await_stable_state(Microtask::ImageElement(task));
+    }
+
+    /// Step 2-12 of https://html.spec.whatwg.org/multipage/#img-environment-changes
+    fn react_to_environment_changes_sync_steps(&self, generation: u32) {
+        fn add_cache_listener_for_element(image_cache: Arc<ImageCache>,
+                                          id: PendingImageId,
+                                          elem: &HTMLImageElement) {
+            let trusted_node = Trusted::new(elem);
+            let (responder_sender, responder_receiver) = ipc::channel().unwrap();
+
+            let window = window_from_node(elem);
+            let task_source = window.networking_task_source();
+            let task_canceller = window.task_canceller(TaskSourceName::Networking);
+            let generation = elem.generation.get();
+            ROUTER.add_route(responder_receiver.to_opaque(), Box::new(move |message| {
+                debug!("Got image {:?}", message);
+                // Return the image via a message to the script thread, which marks
+                // the element as dirty and triggers a reflow.
+                let element = trusted_node.clone();
+                let image = message.to().unwrap();
+                let _ = task_source.queue_with_canceller(
+                    task!(process_image_response_for_environment_change: move || {
+                        let element = element.root();
+                        // Steps 3-4
+                        let (selected_source, selected_pixel_density) = match element.select_image_source() {
```

Comment:
Instead of calling this, we should pass the selected source and density as arguments to add_cache_listener_for_element so we can use them inside this task.